homeassistant/helpers/config_entry_oauth2_flow.py

```diff
@@ -0,0 +1,397 @@
+        self.token_url = token_url
+
+    @property
+    def name(self) -> str:
+        """Name of the implementation."""
+        return "Configuration.yaml"
+
+    @property
+    def domain(self) -> str:
+        """Domain providing the implementation."""
+        return self._domain
+
+    @property
+    def redirect_uri(self) -> str:
+        """Return the redirect uri."""
+        return f"{self.hass.config.api.base_url}{AUTH_CALLBACK_PATH}"  # type: ignore
+
+    async def async_generate_authorize_url(self, flow_id: str) -> str:
+        """Generate a url for the user to authorize."""
+        return str(
+            URL(self.authorize_url).with_query(
+                {
+                    "response_type": "code",
+                    "client_id": self.client_id,
+                    "redirect_uri": self.redirect_uri,
+                    "state": _encode_jwt(self.hass, {"flow_id": flow_id}),
+                }
+            )
+        )
+
+    async def async_resolve_external_data(self, external_data: Any) -> dict:
+        """Resolve the authorization code to tokens."""
+        session = async_get_clientsession(self.hass)
+
+        data = {
+            "grant_type": "authorization_code",
+            "client_id": self.client_id,
+            "code": external_data,
+            "redirect_uri": self.redirect_uri,
+        }
+
+        if self.client_secret is not None:
+            data["client_secret"] = self.client_secret
+
+        resp = await session.post(self.token_url, data=data)
+
+        resp.raise_for_status()
+
+        new_token = await resp.json()
+        new_token["expires_at"] = time.time() + new_token["expires_in"]
+        return cast(dict, new_token)
+
+    async def async_refresh_token(self, token: dict) -> dict:
+        """Refresh tokens."""
+        session = async_get_clientsession(self.hass)
+
+        data = {
+            "grant_type": "refresh_token",
+            "client_id": self.client_id,
+            "refresh_token": token["refresh_token"],
+        }
+
+        if self.client_secret is not None:
+            data["client_secret"] = self.client_secret
+
+        resp = await session.post(self.token_url, json=data)
+
+        resp.raise_for_status()
+
+        new_token = await resp.json()
+        new_token["expires_at"] = time.time() + new_token["expires_in"]
+
+        return {**token, **new_token}
+
+
+class AbstractOAuth2FlowHandler(config_entries.ConfigFlow, metaclass=ABCMeta):
+    """Handle a config flow."""
+
+    DOMAIN = ""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_UNKNOWN
+
+    def __init__(self) -> None:
+        """Instantiate config flow."""
+        if self.DOMAIN == "":
+            raise TypeError(
+                f"Can't instantiate class {self.__class__.__name__} without DOMAIN being set"
+            )
+
+        self.external_data: Any = None
+        self.flow_impl: AbstractOAuth2Implementation = None  # type: ignore
```

Comment:
`Optional[AbstractOAuth2Implementation]` so you can remove `# type: ignore` because python don't know lazy object variable initializing and it is the correct type for python.

Reply:
The reason I didn't do it is because I know it's not None in the places where it's used. Now typing is working against me and I need to add if-checks everywhere.